client/my-sites/plans-features-main/index.tsx

```diff
@@ -509,7 +499,7 @@ const PlansFeaturesMain = ( {
 			displayedIntervals: filteredDisplayedIntervals,
 			showPlanTypeSelectorDropdown,
 			kind: planTypeSelector,
-			plans: gridPlansForFeaturesGrid.map( ( gridPlan ) => gridPlan.planSlug ),
+			plans: gridPlansForFeaturesGrid?.map( ( gridPlan ) => gridPlan.planSlug ) || [],
```

Comment:
Fair question. After a few days away from this diff, I needed to refresh my memory :)

It relates to how the `isLoadingGridPlans` state is derived in `PlansFeaturesMain`. 

Previous:

```ts
const isLoadingGridPlans = Boolean( ! intent || ! gridPlans );
```

Now:

```ts
const isLoadingGridPlans = Boolean(
  ! intent || ! gridPlansForFeaturesGrid || ! gridPlansForComparisonGrid
);
```

The `gridPlansForXxGrid` values are derived from the `useGridPlansForFeaturesGrid` and `useGridPlansForComparisonGrid` hooks, and we have moved the derivation of `gridPlans` (ie. `gridPlans = useGridPlans()`) into those hooks.

As `gridPlans` being `null` indicates that the plans are loading (something I discovered as part of this refactor), we need to maintain that contract and pass `null` back from the hooks. 

```ts
return useMemo( () => {
  if ( ! gridPlans ) {
	return null;
  }

  return gridPlans.reduce()
  ...
```

If instead we just returned `[]` from the hooks, it could mean that the plans have actually loaded and there are no plans.

Reply:
Thanks for outlining with more insight. I totally also prefer returning `null` when possible to indicate an "empty state". I certainly lean more and more on variant uses of `undefined` and `null` to indicate "emptiness", in general (as a principle, and where it does not relate to more complex structures that embody their own state/status mechanism). `undefined` for something that has not been observed, `null` for something that has been observed but not found or emptied thereafter. I believe this is also reflected in Don Abramov's [Just Javascript](https://justjavascript.com/) course, if not directly so.

Now the question is whether it makes sense to be explicitly passing an empty array into one of the grids e.g. 

```
<FeaturesGrid gridPlans={ gridPlans || [] } />
``` 

or conditionally rendering the grid instead perhaps e.g. 

```
{ gridPlans && <FeaturesGrid gridPlans={ gridPlans } /> }
```

IMO the second approach is safer, as developers can better assume they are dealing with a defined set of grid plans, rather than a potentially empty one, but I could be wrong. I think there are a few points to consider (related):

- should the Features/Comparison Grids be responsible for some sort of view "state" (like a placeholder) when an empty array is passed? or should that happen at the consuming end instead? and at what level do we orchestrate that if inside the grids (meaning, we'd probably want the placeholder to be based off a loading status that's a bit longer instead)
- should the grids even have to worry about an "empty array" being passed for grid plans? if so, then how? 
- what's the safest to avoid the possibility of glitches in the UI? e.g. some skeleton renders with the empty array and then expanded to fit plans when array's populated (this would be a glitch in the UI).

So IMO, if the grids have to deal with an "empty state", then it would be best to happen over the longest interval (i.e. the loading status), which is not readily possible. I would probably choose conditionally rendering in that case, or some sort of early/crude null return.

Reply:
Thanks for the feedback. I agree, conditionally rendering the grids is a cleaner approach here. Updated: https://github.com/Automattic/wp-calypso/pull/87478/commits/656ad67078189198b626364fbcb724f104381287

The required `plans` prop for the `PlanTypeSelector` component is a slightly more 'interesting' case, where we build up a `planTypeSelectorProps` object that's shared in a few spots. I've pushed a separate update (https://github.com/Automattic/wp-calypso/pull/87478/commits/8ec714af23bb2f3bf6af3dab6c5f363299267fa1) for this case that uses conditional rendering. This is an interim fix, though, as I think we should refactor all of the 'setup' code for `PlanTypeSelector` out of `plans-features-main` in a followup. Wdyt?

Created an issue for it here.